glue-etl/src/main/scala/com/circle/data/utils/QueryBase.scala

```diff
@@ -86,43 +86,58 @@ object QueryBase {
 
   /**
    * Get SeedInvest investor email preference
+   * We have to do it complex query since the ponyexpress tables don't have the userId.
+   * We can only join on the email address.
    * SELECT
+       profile.entity_ptr_id AS userId
        rule.id,
        rule.created_at,
        rule.modified_at,
        rule.active AS subscribed,
        channel.contact_info AS email,
-       preference.key,
-       preference.label,
-       preference.display_label
+       preference.key
      FROM public_ponyexpress_preference_rule rule
        INNER JOIN public_ponyexpress_preferences_contact_channel channel
          ON rule.contact_channel_id=channel.id
        INNER JOIN public_ponyexpress_preferences_preference preference
          ON rule.preference_id=preference.id
+       INNER JOIN public_auth
+         ON channel.contact_info = auth.email
+       INNER JOIN public_seedinvest_user_userprofile profile
+         ON profile.user_id = auth.id
    */
   def getEmailPreferenceData(
+    authData: DataFrame,
+    profileData: DataFrame,
     preferenceRuleData: DataFrame,
     contactChannelData: DataFrame,
     preferenceData: DataFrame
   ): DataFrame = {
     val selectCols = Array(
+      "profile.entity_ptr_id AS `userId`",
       "rule.id",
       "rule.created_at",
       "rule.modified_at",
       "rule.active AS `subscribed`",
       "channel.contact_info AS `email`",
-      "preference.key",
-      "preference.label",
-      "preference.display_label"
+      "preference.key"
     )
 
     val result = preferenceRuleData.as("rule")
       .join(contactChannelData.as("channel"), col("channel.id") === col("rule.contact_channel_id"))
       .join(preferenceData.as("preference"), col("preference.id") === col("rule.preference_id"))
+      .join(authData.as("auth"), col("auth.email") === col("channel.contact_info"))
+      .join(profileData.as("profile"), col("profile.user_id") === col("auth.id"))
+      .filter(col("active") === "true")
       .selectExpr(selectCols: _*)
       // Make sure the email has right format
       .filter(col("email").rlike("""^([a-zA-Z0-9_\-\.]+)@([a-zA-Z0-9_\-\.]+)\.([a-zA-Z]{2,5})$"""))
```

Comment:
Hmm... Do we have this problem in the other jobs?
Also, email validation/parsing is tricky. Where did you get the regex?

Reply:
A valid email address according to the RFC could contain many other characters that the ones specified in this google search (although the most common ones are listed here).
FredB!oggs@example.com ==> example

What I don't want to happen is we get a real user who doesn't have any email preferences set on them because we are improperly filtering them out here. Can't we just send them all?

Reply:
Good catch. Relax the regex a little bit.